qiskit/transpiler/passes/routing/layout_transformation.py

```diff
@@ -0,0 +1,76 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Map (with minimum effort) a DAGCircuit onto a `coupling_map` adding swap gates."""
+
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+from qiskit.transpiler.routing import util
+from qiskit.transpiler.routing.general import ApproximateTokenSwapper
+
+
+class LayoutTransformation(TransformationPass):
+    """ Adds a Swap circuit for a given (partial) permutation to the circuit.
+
+    This circuit is found by a 4-approximation algorithm for Token Swapping.
+    More details are available in the routing code.
+    """
+
+    def __init__(self, coupling_map, initial_layout, final_layout, trials=4):
+        """LayoutTransformation initializer.
+
+        Args:
+            coupling_map (CouplingMap): Directed graph represented a coupling map.
+            initial_layout (Layout): The starting layout of qubits onto physical qubits.
+            final_layout (Layout): The final layout of qubits on phyiscal qubits.
+            trials (int): How many randomized trials to perform, taking the best circuit as output.
+        """
+        super().__init__()
+        self.coupling_map = coupling_map
+        self.initial_layout = initial_layout
+        self.final_layout = final_layout
+        graph = coupling_map.graph.to_undirected()
+        token_swapper = ApproximateTokenSwapper(graph)
+
+        # Find the permutation that between the initial physical qubits and final physical qubits.
+        permutation = {pqubit: final_layout.get_virtual_bits()[vqubit]
+                       for vqubit, pqubit in initial_layout.get_virtual_bits().items()}
+        swaps = token_swapper.map(permutation, trials)
+        # None of the swaps are guaranteed to be disjoint so we perform one swap every layer.
+        parallel_swaps = [[swap] for swap in swaps]
+        self.permutation_circuit = util.circuit(parallel_swaps)
+
+    def run(self, dag):
+        """Apply the specified partial permutation to the circuit.
+
+        Args:
+            dag (DAGCircuit): DAG to transform the layout of.
+
+        Returns:
+            DAGCircuit: The DAG with transformed layout.
+
+        Raises:
+            TranspilerError: if the coupling map or the layout are not
+            compatible with the DAG.
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get('q', None) is None:
+            raise TranspilerError('LayoutTransform runs on physical circuits only')
+
+        if len(dag.qubits()) > len(self.coupling_map.physical_qubits):
+            raise TranspilerError('The layout does not match the amount of qubits in the DAG')
+
+        edge_map = {vqubit: self.initial_layout.get_physical_bits()[pqubit]
```

Comment:
```
edge_map = {right_q: dag.qubits()[left_q] 
            for (left_q, right_q) in self.permutation_circuit.inputmap.items()}
```
would work to return a physical-to-physical qubit map. (Note: Both of left_q and right_q are physical qubits.)

Reply:
I'm not sure why you can assume the same ordering of the qubits by using `dag.qubits()[left_q]`. Why would the physical qubit have the same index as the qubit register in the dag? Is that guaranteed somehow?

E.g., if I get a Layout with a DagCircuit which say that qubit `v[15] → 35`, then that is guaranteed to mean that virtual qubit v15 is mapped onto dag.qubits()[35]?

Reply:
ApplyLayout guarantees that the physical qubit have the same index as the qubit register in the dag. It assures that the dag has only one register 'q' and its index is the same as that of physical qubits, i.e. the dag never has virtual qubits. So as long as LayoutTransformation runs after ApplyLayout, `dag.qubits()[left_q]` is valid. If we want to handle circuits with virtual qubits (in future), we need more lines here to care the case you suggested.